spec/p2p/v0.34/pex-protocol.md

```diff
@@ -0,0 +1,240 @@
+# Peer Exchange Protocol
+
+The Peer Exchange (PEX) protocol enables nodes to exchange peer addresses, thus
+implementing a peer discovery mechanism.
+
+The PEX protocol uses two messages:
+
+- `PexRequest`: sent by a node to [request](#requesting-addresses) peer
+  addresses to a peer
+- `PexAddrs`: a list of peer addresses [provided](#providing-addresses) to a
+  peer as response to a `PexRequest` message
+
+While all nodes, with few exceptions, participate on the PEX protocol,
+a subset of nodes, configured as [seed nodes](#seed-nodes) have a particular
+role in the protocol.
+They crawl the network, connecting to random peers, in order to learn as many
+peer addresses as possible to provide to other nodes.
+
+## Requesting Addresses
+
+A node requests peer addresses by sending a `PexRequest` message to a peer.
+
+For regular nodes, not operating in seed mode, a PEX request is sent when
+the node *needs* peers addresses, a condition checked:
+
+1. When an *outbound* peer is added, causing the node to request addresses from
+   the new peer
+2. Periodically, by the `ensurePeersRoutine`, causing the node to request peer
+   addresses to a randomly selected peer
+
+A node needs more peer addresses when its addresses book has
+[less than 1000 records](./addressbook.md#need-for-addresses).
+It is thus reasonable to assume that the common case is that a peer needs more
+peer addresses, so that PEX requests are sent whenever the above two situations happen.
+
+A PEX request is sent when a new *outbound* peer is added.
+The same does not happen with new inbound peers
+because outbound peers, that the node has dialed, are considered more trustworthy
+than inbound peers, that the node has accepted.
```

Comment:
What is the rationale to consider them more trustworthy?
Is this distinction (inbound _vs_ outbound) also used for point 2)? If not, then there is a contradiction, IMO, on what is considered trustworthy.

Reply:
Good point, at the moment the `ensurePeersRoutine` does not take into consideration whether the peer in the address book is outbound or inbound but does check the rating of the peers (chooses which bucket to take peer from).  The PEX request in case 1) is sent immediately to this outbound peer assuming that the node already trusted this peer as the outbound peers are usually preconfigured by the node operators.

Reply:
Still, I'd add some footnote or something, as the "trustworthy" concepts looks more of a heuristic than something well defined.